Allow KILL QUERY for backups

vitlibar opened the pull request:
### Changelog category:
- Improvement

### Changelog entry:
Allow `KILL QUERY` to cancel backups / restores. This PR also makes running backups and restores visible in `system.processes`. Also there is a new setting in the server configuration now - `shutdown_wait_backups_and_restores` (default=true) which makes the server either wait on shutdown for all running backups and restores to finish or just cancel them.

vitlibar committed:
Allow KILL QUERY to cancel backups/restores.

vitlibar committed:
Add tests for cancelling backup/restore queries.

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #58804 from vitlibar/ClickHouse:allow-kill-query-for-backups -->
*This is an automated comment for commit 5eafd54676c9acad6703a4085bc19bb68126d264 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58804/5eafd54676c9acad6703a4085bc19bb68126d264/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58804/5eafd54676c9acad6703a4085bc19bb68126d264/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

vitlibar committed:
Use special statuses BACKUP_CANCELLED and RESTORE_CANCELLED in system.backups

vitlibar committed:
Use the same row in system.processes and the same query_id for both a BACKUP/RESTORE query started by user and its asynchronous execution in a separate thread.

vitlibar committed:
Add server settings "shutdown_wait_backups_and_restores" (default=true)
to set whether shutdown should wait for running backups to finish or just cancel them.

vitlibar commented:
@novikd This PR is ready for review.

novikd commented on src/Backups/BackupsWorker.cpp:
```diff
@@ -468,19 +525,24 @@ void BackupsWorker::doBackup(
 {
-    std::optional<CurrentThread::QueryScope> query_scope;
+    SCOPE_EXIT_SAFE(
+        if (called_async)
```
Also check if `thread_group != nullptr`?

novikd reviewed (approved):
In general LGTM

vitlibar commented on src/Backups/BackupEntriesCollector.cpp:
```diff
@@ -629,8 +643,12 @@ void BackupEntriesCollector::lockTablesForReading()
+        if (!storage)
+            continue;
+
+        checkQueryNotCancelled();
```
I'm calling `checkQueryNotCancelled()` before doing anything time-consuming. Checking a pointer for null is not time-consuming.

vitlibar commented on src/Interpreters/ProcessList.cpp:
```diff
@@ -541,6 +558,28 @@ CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id,
+        elem->is_cancelling = true;
+    }
+
+    SCOPE_EXIT({
+        DENY_ALLOCATIONS_IN_SCOPE;
+
+        auto lock = unsafeLock();
+        elem->is_cancelling = false;
+        cancelled_cv.notify_all();
+    });
+
```
It looks more logically consistent now - because first we set `is_cancelling` to true, then we call `cancelQuery(kill)` and set `is_cancelling` to false. Anyway the difference is not quite essential because `safeLock()` should never throw.

vitlibar committed:
Rename function checkQueryNotCancelled() and add check for thread_group.

vitlibar commented on src/Backups/BackupsWorker.cpp:
```diff
@@ -815,14 +904,20 @@ void BackupsWorker::doRestore(
 {
-    std::optional<CurrentThread::QueryScope> query_scope;
+    SCOPE_EXIT_SAFE(
+        if (called_async)
```
done

vitlibar commented on src/Backups/BackupEntriesCollector.h:
```diff
@@ -97,11 +100,15 @@ class BackupEntriesCollector : private boost::noncopyable
     Strings setStage(const String & new_stage, const String & message = "");
 
+    /// Throws an exception if the BACKUP query was cancelled.
+    void checkQueryNotCancelled() const;
```
I renamed it to `checkIsQueryCancelled()`

vitlibar commented:
Test failures:
- [02901_parallel_replicas_rollup](https://s3.amazonaws.com/clickhouse-test-reports/58804/1a1b0894ebf3c35a4ae20a1775ff75a1eeb56202/stateless_tests__release__analyzer_.html) and [02891_array_shingles](https://s3.amazonaws.com/clickhouse-test-reports/58804/1a1b0894ebf3c35a4ae20a1775ff75a1eeb56202/stateless_tests__release__analyzer_.html) are [flaky](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0ZXN0X25hbWUsIGNoZWNrX3N0YXJ0X3RpbWUsIHB1bGxfcmVxdWVzdF9udW1iZXIsIHJlcG9ydF91cmwKZnJvbSBjaGVja3Mgd2hlcmUgJzIwMjQtMDEtMTMnIDw9IGNoZWNrX3N0YXJ0X3RpbWUgYW5kICh0ZXN0X25hbWUgaW4gWycwMjg5MV9hcnJheV9zaGluZ2xlcycsICcwMjkwMV9wYXJhbGxlbF9yZXBsaWNhc19yb2xsdXAnXSkgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIG9yZGVyIGJ5IGNoZWNrX3N0YXJ0X3RpbWU=)
- [00157_cache_dictionary](https://s3.amazonaws.com/clickhouse-test-reports/58804/1a1b0894ebf3c35a4ae20a1775ff75a1eeb56202/stateful_tests__msan__parallelreplicas_.html) and [00082_quantiles](https://s3.amazonaws.com/clickhouse-test-reports/58804/1a1b0894ebf3c35a4ae20a1775ff75a1eeb56202/stateful_tests__msan__parallelreplicas_.html) are [flaky](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0ZXN0X25hbWUsIGNoZWNrX3N0YXJ0X3RpbWUsIHB1bGxfcmVxdWVzdF9udW1iZXIsIHJlcG9ydF91cmwKZnJvbSBjaGVja3Mgd2hlcmUgJzIwMjQtMDEtMTMnIDw9IGNoZWNrX3N0YXJ0X3RpbWUgYW5kICh0ZXN0X25hbWUgaW4gWycwMDA4Ml9xdWFudGlsZXMnLCAnMDAxNTdfY2FjaGVfZGljdGlvbmFyeSddKSBhbmQgdGVzdF9zdGF0dXMgaW4gKCdGQUlMJywgJ0ZMQUtZJykgb3JkZXIgYnkgY2hlY2tfc3RhcnRfdGltZQ==)
